packages/expo-notifications/src/DevicePushTokenAutoRegistration.fx.ts

```diff
@@ -0,0 +1,83 @@
+import ServerRegistrationModule from './ServerRegistrationModule';
+import { addPushTokenListener } from './TokenEmitter';
+import { DevicePushToken } from './Tokens.types';
+import {
+  updatePushTokenAsync,
+  hasPushTokenBeenUpdated,
+  interruptPushTokenUpdates,
+} from './utils/updatePushTokenAsync';
+
+/**
+ * Encapsulates device server registration data
+ */
+export type Registration = {
+  url: string;
+  body: Record<string, any>;
+  pendingDevicePushToken?: DevicePushToken | null;
+};
+
+/**
+ * Sets the last registration information so that the device push token
+ * gets pushed to the given registration endpoint
+ * @param registration Registration endpoint to inform of new tokens
+ */
+export async function setAutoServerRegistrationAsync(
+  registration: Omit<Registration, 'pendingDevicePushToken'>
+) {
+  // We are overwriting registration, so we shouldn't let
+  // any pending request complete.
+  interruptPushTokenUpdates();
+  // Remember the registration information for future token updates.
+  await ServerRegistrationModule.setLastRegistrationInfoAsync?.(JSON.stringify(registration));
+}
+
+/**
+ * Removes last Expo server registration, future device push token
+ * updates won't get sent there anymore.
+ */
+export async function removeAutoServerRegistrationAsync() {
+  // We are removing registration, so we shouldn't let
+  // any pending request complete.
+  interruptPushTokenUpdates();
+  // Do not consider any registration when token updates.
+  await ServerRegistrationModule.setLastRegistrationInfoAsync?.(null);
+}
+
+export async function __handlePersistedRegistrationInfoAsync(
+  lastRegistrationInfo: string | null | undefined
+) {
+  if (!lastRegistrationInfo) {
+    // No last registration info, nothing to do
+    return;
+  }
+  try {
+    const lastRegistration: Registration = JSON.parse(lastRegistrationInfo);
+    // We only want to retry if `hasPushTokenBeenUpdated` is false.
+    // If it were true it means that another call to `updatePushTokenAsync`
+    // has already occured which could only happen from the listener
+    // which has newer information than persisted storage.
+    if (lastRegistration?.pendingDevicePushToken && !hasPushTokenBeenUpdated()) {
```

Comment:
```suggestion
    if (lastRegistration.pendingDevicePushToken && !hasPushTokenBeenUpdated()) {
```
I think this is not needed here - the JSON.parse can't return a null value, can it?

Reply:
So we should change the type of `lastRegistration`.